src/snekmate/auth/AccessControl.vy

```diff
@@ -147,8 +131,6 @@ def __init__():
             the `msg.sender`.
     """
     self._grant_role(DEFAULT_ADMIN_ROLE, msg.sender)
```

Comment:
I think, one point to discuss here is if we want to refactor it so that the user can supply an address for the `DEFAULT_ADMIN_ROLE`. If you use often factory contracts (where `msg.sender` would be the factory contract at deployment) you would need to add an additional line in the implementation to fix it:

```vyper
@deploy
@payable
def __init__(account: address):
    ac.__init__()
    ac._grant_role(DEFAULT_ADMIN_ROLE, account)
```

OTOH, the UX & security overall is better via pointing to the `msg.sender` since for most cases you can ensure that there is someone who really holds the private key for the address (except for the factory / smart contract wallet cases of course)

Reply:
There's a general trend to accept one extra parameter to transfer ownership or assign admin role (solmate always had this in `Ownable`, and OZ too in versions 5+).

`Ownable.vy` transfers ownership to `msg.sender` so i'd keep it consistent with this too https://github.com/pcaversaccio/snekmate/blob/e87e4326dfd67b1c7d913c034685e8930dc6c274/src/snekmate/auth/Ownable.vy#L43 

although it's very common for beginners and devs in general to lock contracts out with this when deploying with factories like CREATE2. im curious if there's any data to back a decision here (i.e. how many contract are constructed with an extra parameter vs `msg.sender`)

Reply:
it should be pretty easy for devs to inherit `AcccessControl` and add two lines that grant admin to a constructor param and revoke the msg.sender? if yes, i'm leaning towards using `msg.sender` in `AccessControl`

Reply:
> im curious if there's any data to back a decision here (i.e. how many contract are constructed with an extra parameter vs msg.sender)

Do you have some internal data at OZ maybe on this one? Very curious to see as well.

> it should be pretty easy for devs to inherit AcccessControl and add two lines that grant admin to a constructor param and revoke the msg.sender? if yes, i'm leaning towards using msg.sender in AccessControl

So one issue I see here is that it might not be obvious to do. Let's step through a newcomer:

1. He imports the `AccessControl` module and initialises it.
2. He compiles the contract and the compiler throws with: 
```console
vyper.exceptions.InitializerException: not initialized!
  (hint: add `ac.__init__()` to your `__init__()` function)
```
3. He adds `ac.__init__()` to the constructor accordingly and it compiles and thinks LFG.
4. He deploys via my `CreateX` factory and RIP.

@charles-cooper, call me crazy, but the best solution for this issue would be having constructor overloading I think:

```vy
@deploy
def __init__(admin: address=msg.sender):
    self._grant_role(DEFAULT_ADMIN_ROLE, admin)
```

Unfortunately, the compiler disallows it currently. Thoughts?

Reply:
we don't allow `__init__()` overloading because it doesn't make sense when the constructor is actually called as initcode. i think the best thing from the language perspective would be to implement https://github.com/vyperlang/vyper/issues/3740, then you could have like "factory" creation functions that can be called besides `__init__()`.

Reply:
Hmm, yeah, I see the benefit of the `@deploy` decorator, since you could do something like:

```vy
# AccessControl.vy

@deploy
def regular_init():
    self._grant_role(DEFAULT_ADMIN_ROLE, msg.sender)

@deploy
def init_with_owner(owner: address):
    self._grant_role(DEFAULT_ADMIN_ROLE, owner)
```

And after that the user could do:

```vy
from snekmate.auth import AccessControl as ac
initializes: ac

@deploy
def __init__():
    ac.regular_init()

# OR

@deploy
def __init__(owner: address):
    ac.init_with_owner(owner)
```

I still think tho having something like:

```vy
# AccessControl.vy

@deploy
def init_ac(owner: address=msg.sender):
    self._grant_role(DEFAULT_ADMIN_ROLE, owner)

# Main.vy
from snekmate.auth import AccessControl as ac
initializes: ac

@deploy
def __init__():
    ac.init_ac()

# OR

@deploy
def __init__(owner: address):
    ac.init_ac(owner)
```

would be useful as a language feature.

Reply:
So the decision is to use the yet-to-be-implemented Vyper `@deploy` decorator (https://github.com/vyperlang/vyper/issues/3740) to implement this feature. This might happen with Vyper `0.4.1` however. For this release, we keep the logic as-is.